modules/magniteAnalyticsAdapter.js

```diff
@@ -0,0 +1,885 @@
+    bid.conversionError = true;
+    bid.ogCurrency = currency;
+    bid.ogPrice = cpm;
+    return 0;
+  }
+}
+
+export const parseBidResponse = (bid, previousBidResponse) => {
+  // The current bidResponse for this matching requestId/bidRequestId
+  let responsePrice = getBidPrice(bid)
+  // we need to compare it with the previous one (if there was one) log highest only
+  // THIS WILL CHANGE WITH ALLOWING MULTIBID BETTER
+  if (previousBidResponse && previousBidResponse.bidPriceUSD > responsePrice) {
+    return previousBidResponse;
+  }
+
+  return pick(bid, [
+    'bidPriceUSD', () => responsePrice,
+    'dealId', dealId => dealId || undefined,
+    'mediaType',
+    'dimensions', () => {
+      const width = bid.width || bid.playerWidth;
+      const height = bid.height || bid.playerHeight;
+      return (width && height) ? { width, height } : undefined;
+    },
+    'floorValue', () => deepAccess(bid, 'floorData.floorValue'),
+    'floorRuleValue', () => deepAccess(bid, 'floorData.floorRuleValue'),
+    'floorRule', () => debugTurnedOn() ? deepAccess(bid, 'floorData.floorRule') : undefined,
+    'adomains', () => {
+      const adomains = deepAccess(bid, 'meta.advertiserDomains');
+      const validAdomains = Array.isArray(adomains) && adomains.filter(domain => typeof domain === 'string');
+      return validAdomains && validAdomains.length > 0 ? validAdomains.slice(0, 10) : undefined
+    },
+    'conversionError', conversionError => conversionError === true || undefined, // only pass if exactly true
+    'ogCurrency',
+    'ogPrice'
+  ]);
+}
+
+const addFloorData = floorData => {
+  if (floorData.location === 'noData') {
+    return pick(floorData, [
+      'location',
+      'fetchStatus',
+      'floorProvider as provider'
+    ]);
+  } else {
+    return pick(floorData, [
+      'location',
+      'modelVersion as modelName',
+      'modelWeight',
+      'modelTimestamp',
+      'skipped',
+      'enforcement', () => deepAccess(floorData, 'enforcements.enforceJS'),
+      'dealsEnforced', () => deepAccess(floorData, 'enforcements.floorDeals'),
+      'skipRate',
+      'fetchStatus',
+      'floorMin',
+      'floorProvider as provider'
+    ]);
+  }
+}
+
+let pageReferer;
+
+const getTopLevelDetails = () => {
+  let payload = {
+    channel: 'web',
+    integration: rubiConf.int_type || DEFAULT_INTEGRATION,
+    referrerUri: pageReferer,
+    version: '$prebid.version$',
+    referrerHostname: magniteAdapter.referrerHostname || getHostNameFromReferer(pageReferer),
+    timestamps: {
+      timeSincePageLoad: performance.now(),
+      eventTime: Date.now(),
+      prebidLoaded: magniteAdapter.MODULE_INITIALIZED_TIME
+    }
+  }
+
+  // Add DM wrapper details
+  if (rubiConf.wrapperName) {
+    payload.wrapper = {
+      name: rubiConf.wrapperName,
+      family: rubiConf.wrapperFamily,
+      rule: rubiConf.rule_name
+    }
+  }
+
+  // Add session info
+  const sessionData = storage.localStorageIsEnabled() && updateRpaCookie();
```

Comment:
I expect the method `getTopLevelDetails` to only read data and not write anything; but then inside I find this `updateRpaCookie()` that fires a side effect and I think this should be moved out of here.

Reply:
Good catch I was testing you 😛.

I have made the update.